env0/resource_environment.go

```diff
@@ -277,43 +278,97 @@ func setEnvironmentSchema(d *schema.ResourceData, environment client.Environment
 	return nil
 }
 
+func createVariable(configurationVariable *client.ConfigurationVariable) interface{} {
+	variable := make(map[string]interface{})
+
+	variable["name"] = configurationVariable.Name
+	variable["value"] = configurationVariable.Value
+
+	if configurationVariable.Type == nil || *configurationVariable.Type == 0 {
+		variable["type"] = "environment"
+	} else {
+		variable["type"] = "terraform"
+	}
+
+	if configurationVariable.Description != "" {
+		variable["description"] = configurationVariable.Description
+	}
+
+	if configurationVariable.Regex != "" {
+		variable["regex"] = configurationVariable.Regex
+	}
+
+	if configurationVariable.IsSensitive != nil {
+		variable["is_sensitive"] = configurationVariable.IsSensitive
+	}
+
+	if configurationVariable.IsReadOnly != nil {
+		variable["is_read_only"] = configurationVariable.IsReadOnly
+	}
+
+	if configurationVariable.IsRequired != nil {
+		variable["is_required"] = configurationVariable.IsRequired
+	}
+
+	if configurationVariable.Schema != nil {
+		variable["schema_type"] = configurationVariable.Schema.Type
+		variable["schema_enum"] = configurationVariable.Schema.Enum
+		variable["schema_format"] = configurationVariable.Schema.Format
+	}
+
+	return variable
+}
+
 func setEnvironmentConfigurationSchema(d *schema.ResourceData, configurationVariables []client.ConfigurationVariable) {
-	var variables []interface{}
+	ivariables, ok := d.GetOk("configuration")
+	if !ok {
+		return
+	}
 
-	for _, configurationVariable := range configurationVariables {
-		variable := make(map[string]interface{})
-		variable["name"] = configurationVariable.Name
-		variable["value"] = configurationVariable.Value
-		if configurationVariable.Type == nil || *configurationVariable.Type == 0 {
-			variable["type"] = "environment"
-		} else {
-			variable["type"] = "terraform"
-		}
-		if configurationVariable.Description != "" {
-			variable["description"] = configurationVariable.Description
-		}
-		if configurationVariable.Regex != "" {
-			variable["regex"] = configurationVariable.Regex
-		}
-		if configurationVariable.IsSensitive != nil {
-			variable["is_sensitive"] = configurationVariable.IsSensitive
-		}
-		if configurationVariable.IsReadOnly != nil {
-			variable["is_read_only"] = configurationVariable.IsReadOnly
+	if ivariables == nil {
+		ivariables = make([]interface{}, 0)
+	}
+
+	variables := ivariables.([]interface{})
+
+	newVariables := make([]interface{}, 0)
+
+	// The goal is to maintain existing state order as much as possible. (The backend response order may vary from state).
```

Comment:
comments explain it. Also you can see it in the unit test.
The order of the state and the order returned from the API may vary.

To avoid drifts we try to keep the order as described in the state.